ports/esp32/modnetwork.c

```diff
@@ -454,45 +462,80 @@ STATIC mp_obj_t esp_status(size_t n_args, const mp_obj_t *args) {
+        t->items[0] = mp_obj_new_bytes(wifi_ap_records[i].ssid, ssid_len);
+        t->items[1] = mp_obj_new_bytes(wifi_ap_records[i].bssid, sizeof(wifi_ap_records[i].bssid));
+        t->items[2] = MP_OBJ_NEW_SMALL_INT(wifi_ap_records[i].primary);
+        t->items[3] = MP_OBJ_NEW_SMALL_INT(wifi_ap_records[i].rssi);
+        t->items[4] = MP_OBJ_NEW_SMALL_INT(wifi_ap_records[i].authmode);
+        t->items[5] = mp_const_false; // XXX hidden?
+        mp_obj_list_append(list, MP_OBJ_FROM_PTR(t));
+    }
+    free(wifi_ap_records);
+    wifi_sta_scan_done = false;
+    return list;
+}
+
+STATIC MP_DEFINE_CONST_FUN_OBJ_1(esp_get_scan_results_obj, esp_get_scan_results);
+
+STATIC mp_obj_t esp_scan(size_t n_args, const mp_obj_t *args) {
+    mp_obj_t *self_in = MP_OBJ_TO_PTR(args[0]);
+    bool blocking = true;
+
+    // If 1 arg or 2nd arg is True, blocking scan
+    if (n_args > 1) {
+        blocking = mp_obj_is_true(args[1]);
+    }
+
     // check that STA mode is active
     wifi_mode_t mode;
     ESP_EXCEPTIONS(esp_wifi_get_mode(&mode));
     if ((mode & WIFI_MODE_STA) == 0) {
         mp_raise_msg(&mp_type_OSError, MP_ERROR_TEXT("STA must be active"));
     }
 
-    mp_obj_t list = mp_obj_new_list(0, NULL);
     wifi_scan_config_t config = { 0 };
     // XXX how do we scan hidden APs (and if we can scan them, are they really hidden?)
-    MP_THREAD_GIL_EXIT();
-    esp_err_t status = esp_wifi_scan_start(&config, 1);
-    MP_THREAD_GIL_ENTER();
-    if (status == 0) {
-        uint16_t count = 0;
-        ESP_EXCEPTIONS(esp_wifi_scan_get_ap_num(&count));
-        wifi_ap_record_t *wifi_ap_records = calloc(count, sizeof(wifi_ap_record_t));
-        ESP_EXCEPTIONS(esp_wifi_scan_get_ap_records(&count, wifi_ap_records));
-        for (uint16_t i = 0; i < count; i++) {
-            mp_obj_tuple_t *t = mp_obj_new_tuple(6, NULL);
-            uint8_t *x = memchr(wifi_ap_records[i].ssid, 0, sizeof(wifi_ap_records[i].ssid));
-            int ssid_len = x ? x - wifi_ap_records[i].ssid : sizeof(wifi_ap_records[i].ssid);
-            t->items[0] = mp_obj_new_bytes(wifi_ap_records[i].ssid, ssid_len);
-            t->items[1] = mp_obj_new_bytes(wifi_ap_records[i].bssid, sizeof(wifi_ap_records[i].bssid));
-            t->items[2] = MP_OBJ_NEW_SMALL_INT(wifi_ap_records[i].primary);
-            t->items[3] = MP_OBJ_NEW_SMALL_INT(wifi_ap_records[i].rssi);
-            t->items[4] = MP_OBJ_NEW_SMALL_INT(wifi_ap_records[i].authmode);
-            t->items[5] = mp_const_false; // XXX hidden?
-            mp_obj_list_append(list, MP_OBJ_FROM_PTR(t));
+
+    if (blocking) {
+        MP_THREAD_GIL_EXIT();
+        esp_err_t status = esp_wifi_scan_start(&config, 1);
+        MP_THREAD_GIL_ENTER();
+
+        if (status == 0) {
+            return esp_get_scan_results(self_in);
         }
-        free(wifi_ap_records);
+    } else {
+        ESP_LOGI("wifi", "Attempting non-blocking scan");
+        ESP_EXCEPTIONS(esp_wifi_scan_start(&config, 0));
```

Comment:
I know it had `1` there before, but the `block` param is actually a bool. Would be clearer to use `true`/`false`.

`esp_err_t esp_wifi_scan_start(const wifi_scan_config_t *config, bool block);`

Maybe for a different PR but I notice the two paths have different exception handling (in the original code there was no handling for `status != 0`. Might as well make both paths use `ESP_EXCEPTIONS`.

Maybe

```
wifi_sta_scan_done = false;

MP_THREAD_GIL_EXIT();
esp_err_t status =esp_wifi_scan_start(&config, blocking);
MP_THREAD_GIL_ENTER();
ESP_EXCEPTIONS(status);

if (blocking) {
  return esp_get_scan_results(self_in);
} else {
  return mp_const_none;
}

Reply:
Agree with both points, the exception handling change is pretty small and more robust, happy to include it, let me know it it really should be out own PR.